File: src/loaders/excel_converter.py
"""
Conversor de XLSX para CSV.
Suporta multiplos sheets, usa polars para leitura eficiente.
"""

import os
import pandas as pd
import polars as pl


class ExcelConverter:
    """Converte arquivos Excel (.xlsx) para CSV."""

    def __init__(self, logger=None):
        self.logger = logger

    def _log(self, msg: str):
        if self.logger:
            self.logger.info(msg)
        else:
            print(f"[CONVERSOR] {msg}")

    def listar_sheets(self, caminho_xlsx: str) -> list:
        """Lista todas as abas (sheets) do arquivo Excel."""
        try:
            xl = pd.ExcelFile(caminho_xlsx)
            sheets = xl.sheet_names
            self._log(f"Sheets encontradas: {sheets}")
            return sheets
        except Exception as e:
            self._log(f"ERRO ao ler abas com openpyxl: {e}")
            self._log("Tentando com polars...")
            try:
                xf = pl.ExcelWorkbook(caminho_xlsx)
                sheets = xf.sheet_names()
                self._log(f"Sheets encontradas (polars): {sheets}")
                return sheets
            except Exception as e2:
                self._log(f"ERRO ao ler abas com polars: {e2}")
                return []

    def converter(self, caminho_xlsx: str, pasta_saida: str,
                  sheet: str = 0, nome_saida: str = None,
                  casas_decimais: int = None) -> str:
        """
        Converte uma aba do Excel para CSV.
        Tenta polars primeiro (eficiente), fallback para pandas/openpyxl.
        """
        if not os.path.exists(caminho_xlsx):
            raise FileNotFoundError(f"Arquivo nao encontrado: {caminho_xlsx}")

        os.makedirs(pasta_saida, exist_ok=True)
        self._log(f"Lendo aba '{sheet}' de '{os.path.basename(caminho_xlsx)}'...")

        try:
            df = self._ler_polars(caminho_xlsx, sheet)
        except Exception as e:
            self._log(f"Polars falhou ({e}), tentando pandas...", "warning" if hasattr(self.logger, 'warning') else None)
            df = pd.read_excel(caminho_xlsx, sheet_name=sheet, engine="openpyxl")

        if casas_decimais is not None:
            for col in df.select_dtypes(include=["float"]).columns:
                df[col] = df[col].round(casas_decimais)

        if nome_saida is None:
            base = os.path.splitext(os.path.basename(caminho_xlsx))[0]
            nome_saida = f"{base}_convertido"

        caminho_csv = os.path.join(pasta_saida, f"{nome_saida}.csv")
        df.to_csv(caminho_csv, index=False, encoding="utf-8-sig")

        self._log(f"Convertido: {len(df)} linhas, {len(df.columns)} colunas -> {caminho_csv}")
        return caminho_csv

    def _ler_polars(self, caminho_xlsx: str, sheet) -> pd.DataFrame:
        """Le Excel com polars (calamine engine, extremamente rapido)."""
       xf = pl.ExcelWorkbook(caminho_xlsx)

        if isinstance(sheet, int):
            sheets = xf.sheet_names()
            if sheet < len(sheet):
                sheet_name = sheets[sheet]
            else:
                sheet_name = sheets[0]
        else:
            sheet_name = sheet

        df_pl = xf.read_excel(sheet_name=sheet_name)
        return df_pl.to_pandas()

    def converter_todas_sheets(self, caminho_xlsx: str, pasta_saida: str) -> list:
        """Converte todas as abas do Excel para CSVs separados."""
        sheets = self.listar_sheets(caminho_xlsx)
        caminhos = []
        for sheet in sheets:
            try:
                caminho = self.converter(caminho_xlsx, pasta_saida, sheet=sheet)
                caminhos.append(caminho)
            except Exception as e:
                self._log(f"ERRO ao converter aba '{sheet}': {e}")
        return caminhos

    def detectar_problemas(self, caminho_xlsx: str, sheet: str = 0) -> dict:
        """Detecta problemas comuns antes da conversao."""
        try:
            df = self._ler_polars(caminho_xlsx, sheet)
            if len(df) > 100:
                df = df.head(100)
            problemas = []

            colunas_unnamed = [c for c in df.columns if str(c).startswith("Unnamed")]
            if colunas_unnamed:
                problemas.append(f"Colunas 'Unnamed' detectadas: {colunas_unnamed}")

            nulos_por_col = df.isna().sum()
            colinas_muito_nulas = nulos_por_col[nulos_por_col > len(df) * 0.5].index.tolist()
            if colinas_muito_nulas:
                problemas.append(f"Colunas com muitos nulos: {colinas_muito_nulas}")

            duplicatas = df.duplicated().sum()
            if duplicatas > 0:
                problemas.append(f"{duplicatas} linhas duplicadas (na amostra)")

            return {"tem_problemas": len(problemas) > 0, "problemas": problemas}
        except Exception as e:
            return {"tem_problemas": True, "problemas": [f"Erro ao analisar: {e}"]}

File: src/loaders/excelConverter.ts
import fs from "fs";
import path from "path";
import * as XLSX from "xlsx";
import Papa from "papaparse";
import { DataFrame, DataRow } from "../types";
import { Logger } from "../utils/logger";

export class ExcelConverter {
  private logger?: Logger;

  constructor(logger?: Logger) {
    this.logger = logger;
  }

  private _log(msg: string) {
    if (this.logger) {
      this.logger.info(msg);
    } else {
      console.log(`[CONVERSOR] ${msg}`);
    }
  }

  public listarSheets(caminhoXlsx: string): string[] {
    try {
      const workbook = XLSX.readFile(caminhoXlsx);
      const sheets = workbook.SheetNames;
      this._log(`Sheets encontradas: ${sheets.join(", ")}`);
      return sheets;
    } catch (e: any) {
      this._log(`ERRO ao ler abas: ${e.message}`);
      return [];
    }
  }

  public lerSheet(caminhoXlsx: string, sheetNameOrIndex: string | number = 0): DataFrame {
    if (!fs.existsSync(caminhoXlsx)) {
      throw new Error(`Arquivo não encontrado: ${caminhoXlsx}`);
    }

    const workbook = XLSX.readFile(caminhoXlsx);
    let sheetName: string;

    if (typeof sheetNameOrIndex === "number") {
      sheetName = workbook.SheetNames[sheetNameOrIndex] || workbook.SheetNames[0];
    } else {
      sheetName = sheetNameOrIndex;
    }

    const worksheet = workbook.Sheets[sheetName];
    if (!worksheet) {
      throw new Error(`Aba '${sheetName}' não encontrada.`);
    }

    // Convert sheet to json rows
    const rows = XLSX.utils.sheet_to_json<Record<string, any>>(worksheet, { defval: null, raw: false });

    // Determine columns from rows or range
    const colSet = new Set<string>();
    for (const row of rows) {
      for (const key of Object.keys(row)) {
        colSet.add(key.trim());
      }
    }
    const columns = Array.from(colSet);

    const data: DataRow[] = rows.map((row) => {
      const cleanRow: DataRow = {};
      for (const col of columns) {
        let val = row[col] !== undefined ? row[col] : null;
        if (val === "" || val === undefined) val = null;
        cleanRow[col] = val;
      }
      return cleanRow;
    });

    this._log(`Lida aba '${sheetName}': ${data.length} linhas, ${columns.length} colunas.`);
    return { columns, data };
  }

  public converter(
    caminhoXlsx: string,
    pastaSaida: string,
    sheet: string | number = 0,
    nomeSaida?: string,
    casasDecimais?: number
  ): string {
    const df = this.lerSheet(caminhoXlsx, sheet);

    if (casasDecimais !== undefined) {
      for (const row of df.data) {
        for (const col of df.columns) {
          if (typeof row[col] === "number") {
            row[col] = Number(row[col].toFixed(casasDecimais));
          }
        }
      }
    }

    if (!nomeSaida) {
      const base = path.basename(caminhoXlsx, path.extname(caminhoXlsx));
      nomeSaida = `${base}_convertido`;
    }

    fs.mkdirSync(pastaSaida, { recursive: true });
    const caminhoCsv = path.join(pastaSaida, `${nomeSaida}.csv`);

    const csvContent = Papa.unparse(df.data, {
      columns: df.columns,
      header: true,
    });

    // Write with UTF-8 BOM
    fs.writeFileSync(caminhoCsv, "\uFEFF" + csvContent, { encoding: "utf8" });
    this._log(`Convertido: ${df.data.length} linhas, ${df.columns.length} colunas -> ${caminhoCsv}`);
    return caminhoCsv;
  }

  public detectarProblemas(caminhoXlsx: string, sheet: string | number = 0): { tem_problemas: boolean; problemas: string[] } {
    try {
      const df = this.lerSheet(caminhoXlsx, sheet);
      const sample = df.data.slice(0, 100);
      const problemas: string[] = [];

      const colunasUnnamed = df.columns.filter((c) => c.startsWith("Unnamed") || c.startsWith("__EMPTY"));
      if (colunasUnnamed.length > 0) {
        problemas.push(`Colunas 'Unnamed' detectadas: ${colunasUnnamed.join(", ")}`);
      }

      for (const col of df.columns) {
        const nulos = sample.filter((r) => r[col] === null || r[col] === undefined || r[col] === "").length;
        if (nulos > sample.length * 0.5) {
          problemas.push(`Coluna '${col}' com muitos nulos (${Math.round((nulos / sample.length) * 100)}%)`);
        }
      }

      return {
        tem_problemas: problemas.length > 0,
        problemas,
      };
    } catch (e: any) {
      return {
        tem_problemas: true,
        problemas: [`Erro ao analisar: ${e.message}`],
      };
    }
  }
}
